database/tbcd/level/level.go

```diff
@@ -377,14 +443,33 @@ func (l *ldb) BlockHeaderGenesisInsert(ctx context.Context, wbh *wire.BlockHeade
 	bmBatch := new(leveldb.Batch)
 	bhBatch := new(leveldb.Batch)
 
-	hhKey := heightHashToKey(0, bhash[:])
+	// Genesis insert can be called with an effective genesis block at a particular height
+	// and with a particular cumulative difficulty which is guaranteed canonical (protocol
+	// assumes this effective genesis block will never fork), allowing a header-only TBC
+	// instance to maintain Bitcoin consensus starting from a non-genesis block rather
+	// than requiring all historical state which is irrelevant to conesnsus assuming the
+	// effective genesis block is never forked.
+	hhKey := heightHashToKey(height, bhash[:])
 	hhBatch.Put(hhKey, []byte{})
-	cdiff := big.NewInt(0)
-	cdiff = new(big.Int).Add(cdiff, blockchain.CalcWork(wbh.Bits))
-	ebh := encodeBlockHeader(0, h2b(wbh), cdiff)
+
+	// Handle the default case where the passed-in block is actually the genesis
+	cdiff := blockchain.CalcWork(wbh.Bits)
+
+	// If an effective starting difficulty is supplied and is not set to zero, then use it
+	// instead (used for external header mode). In the event that an effective genesis block
+	// is supplied but the cumulative difficulty is not set, the difficulty of that
+	// effective genesis block (set above) will be retained, which has no effect on local
+	// consensus determination but will not permit direct comparison of cumulative difficulty
+	// against the full chain, only relative comparisons between cumulative difficulties of
+	// blocks on top of the same effective genesis block are guaranteed valid.
+	if diff != nil && diff.Cmp(big.NewInt(0)) > 0 {
+		cdiff = diff
+	}
+	ebh := encodeBlockHeader(height, h2b(wbh), cdiff)
```

Comment:
This should be removed. If we force sending in genesis then crash if diff is nil.